### DISCUSSION TITLE
Welcome to Fabric Discussions!

### DISCUSSION THREAD

**Discussion by danielmiessler** [danielmiessler/Fabric#45] [Announcements]
## 👋 Greetings!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Discuss the current project.
  * Share ideas about possibilities for Fabric.

To start, give us a sentence or two on yourself and let us know your idea!

**Comment by chrissonntag** [comment]


**Reply by CelebritySoulja** [reply]
I'm new here and wants to know more

**Reply by cbruyndoncx** [reply]
Fyi @chrissonntag I just saw that ollama now has openapi compatible interface, I also saw there is a generic client now wich might cover this use case ...

**Reply by waldo323** [reply]
I look forward to using fabric with local models, whether ollama, litellm or whatever fabic ends up working with it.

**Reply by WHITEHACK120** [reply]
hi am john am new here i wanna learn about fabic

**Reply by quiet-ranger** [reply]
Hi! I am JJ (sorry, real name unpronounceable in English) otherwise known as "quiet ranger".  I am an old timer back end developer, Linux tinkler with a penchant for trying new things every so often.  I run into a few issues with fabric but I will post those separately

**Comment by CelebritySoulja** [comment]
Nice and happy to be here

**Comment by Mohammad-gif** [comment]
Explain the background of foundation

**Reply by Dilara760** [reply]
Hi there, I’m Dilara 👋 It's so nice to chat with you. I hope your day is off to a great start. 🌅 What's new in your world? 😄

**Comment by dfinke** [comment]
Hello @danielmiessler. I'm [Doug Finke](https://github.com/dfinke) 15x Microsoft MVP. This project is great. I created the [PowerShell AI Assistant](https://github.com/dfinke/PowerShellAIAssistant) when OpenAI released the new API in November 2023. It mirrors the Python SDK. I also ported an earlier Python SDK version from Jan 2023 [PowerShell AI](https://github.com/dfinke/PowerShellAI) supporting `Text completion` and the early `Chat` approach.

I saw you demo `Fabric` which inspired me to create [PSWeave](https://github.com/dfinke/PSWeave). I borrowed your "system prompts" and added more like: `CompareAndContrast`, `DateBuddy`, `MsGraphWizard`, `PostToX`, `PowerShellMenuWizard`, `PsRegexWizard`.

It leverages the object flow engine in PowerShell that enables piping from one "prompt" to the next, like yours. 
It also uses the new AI Assistant OpenAI API primitives - Assistant, Threads, Messages and Runs. It also all runs in the OpenAI backend.

Would love to chat. Awesome work! - Doug

**Comment by th3f001** [comment]
Hello and wow! this is 100% cool stuff!

I'm Andy, I work in a CyberSec team as Vulnerability Manager, AppSec tester and webapp/API security analyst.

I have a general question: how do you add new patterns? 
i.e. I've drafted my system.md file and added a blank user.md file under `fabric/patterns/MyNewPattern` folder, however `fabric --list` seems to ignore it. 
Is there any specific command that must be given to add new patterns once created?

Thanks and congrats...this is HUGE.
AdV

**Reply by th3f001** [reply]
Oops! It was THAT easy! Apologies for the oversight.
As for the new patterns....any channel where I can submit my eventual patterns before calling for the Pull Request or are you reviewing any new pattern at that stage only?

**Reply by travisdittmer** [reply]
@th3f001 having the same issue with new patterns being ignored by fabric --list. how did you resolve this?

**Reply by travisdittmer** [reply]
this solved the issue for me #167. works great!

**Comment by protteveel** [comment]
Hello, I'm Percy from Redwood City, CA.
I build and lead a PS organization delivering services related to Application Security Testing (AST) tools.
I use ChatGPT regularly. It has increased my performance by roughly 40%.
I imagine, using fabric it will be boosted even more!
Thanks, @danielmiessler! 🙏🏻

**Comment by Luis-Surec** [comment]
Saludos.
Soy nuevo en esto de la programación , estoy empezando con html, javascrip y css.
parecerá  muy básico pero  con el  el evento onclick, solo me funciona una vez, pueden apoyarme les agradezco, dejo mi código abajo
Saludos.

<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.2/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-EVSTQN3/azprG1Anm3QDgpJLIm9Nao0Yz1ztcQTwFspd3yD65VohhpuuCOmLASjC" crossorigin="anonymous">
   
    <link rel="stylesheet" href="../script/encriptador.css">
    <title>encriptador</title>
</head>
<body>
 
<div class="container fondoP">
<div class="container row fondo2"> 
<div class="col-5">
    <textarea  class=" fondo3"name="" id="encriptarTexto" cols="30" rows="10"></textarea>
</div>

<div  class="col-5">
    <textarea class="fondo3"name="" id="recibeTexto" cols="30" rows="10"></textarea>
</div>


<div  class="c-button">
    <button  type="submit"class="btn btn-primary"  onclick="resta()">Encriptar</button>
    <button class="btn btn-success"onclick="">Desencriptar</button>

</div>
</div>


</div>
<script src="../script/encriptador.js"></script>
<script src="https://cdn.jsdelivr.net/npm/bootstrap@5.0.2/dist/js/bootstrap.bundle.min.js" integrity="sha384-MrcW6ZMFYlzcLA8Nl+NtUVF0sA7MsXsP1UyJoMp4YLEuNSfAP+JcXn/tWtIaxVXM" crossorigin="anonymous"></script>
    
</body>
</html>


javascript------------


let  recibeDatos=document.getElementById("encriptarTexto").value;
let resultado = sumar();

function sumar(){
    let valor1=50;
    let valor2=2;
    let resultado= valor1+valor2;
return resultado;
}


function resta(){
let resultadoSuma = sumar();
let restante = resultadoSuma-10;
let enviaDato=document.getElementById("recibeTexto").innerHTML=restante+recibeDatos;

return restante;
}

**Comment by gorbash1370** [comment]
Hello! Firstly, @danielmiessler thank you so much for your generosity sharing your incredible prompts and opening up your workflow. I've gained so much from your UL newsletter and have frequently been inspired by your work ethos, and approach to life in general. 

Reading some of the pull-request dialogue, I noticed mention of a Volunteer list. I'd love to put my hand up for something like this, but... 

I've only been hobby coding for 6 months-ish and I've never worked in tech (career switch in the planning), so regrettably I probably won't be able to offer much to the project at this stage.

(I've actually only just published my first Repo [whisper_wrapper](https://github.com/gorbash1370/whisper_wrapper) - as it happens, an implementation of Whisper for batch transcribing audio files. In novice Python😜 I developed it to process an entire podcast series, for subsequent feeding into an LLM alongside my specialised adaptation of extract_wisdom!)

However, I might be okay for testing - particularly when you're at the stage of gauging suitability for people who are really interested in the project but are less technically advanced.  I'm also pretty good at documentation.  

If there's anything /fabric needs a noob for, I'd love to pitch in 😊

**Comment by xssdoctor** [comment]
Hey thank you! Take a look at a branch I made called transcribe. It does something similar. Let me know if yours is better and we can work on it together

**Comment by Dilara760** [comment]
Hi there, I’m Dilara 👋 It's so nice to chat with you. I hope your day is off to a great start. 🌅 What's new in your world? 😄

**Reply by Luis-Surec** [reply]
Hello hello... Luis is my name, well starting out in the world of programming quite well, nice to greet you, using a translator, hehehe

**Comment by Dilara760** [comment]
You're welcome Luis can we connect to Facebook or Instagram for proper
conversation

**Comment by ValiantEagleSec** [comment]
Do we have to have GPT 4 to use Fabric?

**Reply by nic0711** [reply]
No.
You need either an OpenAI or CLAUDE API key or a local LLM (e.g. Mistral).

**Reply by ValiantEagleSec** [reply]
Dang - I was hoping that was my issue. I keep getting this error when trying to run a pattern, any ideas how to resolve it?

Error code: 404 - {'error': {'message': 'The model `gpt-4-turbo-preview` does not exist or you do not have access to it. Learn more: https://help.openai.com/en/articles/7102672-how-can-i-access-gpt-4.', 'type': 'invalid_request_error', 'param': None, 'code': 'model_not_found'}}

**Reply by xssdoctor** [reply]
make sure you are on the latest version of the repo. so git pull first. then run setup.sh again. If you are on the latest version, run fabric --clear. that error message should have been taken care of in my latest push, that error message used to occur when you set another persistent model then ran fabric --update. I created fabric --clear to get rid of it. Let me know if you still have the issue

So basically, if you are on the latest version of the repo, run fabric --clear. If not, get the latest version and you should be ok

**Reply by ValiantEagleSec** [reply]
Ok so I still have the issue - I git pulled, then ran --clear, then ran setup.sh. Restarted my terminal multiple times. I have tried these two commands, they gave me these errors:

"pbpaste | fabric --pattern summarize"
>Same as the original issue.
"https://getpocket.com/explore/item/the-wizard-of-oz-five-appalling-on-set-stories | fabric --pattern summarize"
>bash: https://getpocket.com/explore/item/the-wizard-of-oz-five-appalling-on-set-stories: No such file or directory
>Same as original issue.

I am new to linux and github, so apologies if my ignorance is making this more complicated than it needs to be, but any help is appreciated.

**Reply by xssdoctor** [reply]
You’re just pasting the paste command into your output. That’s the problem

**Reply by ValiantEagleSec** [reply]
I just realized the way I formatted that was weird. What about for my command that has the link being piped into the pattern?

**Reply by ValiantEagleSec** [reply]
I'm on Ubuntu Linux, here's what I just tried and the result:

"myuser"@Ubuntu:~/Documents$ touch wizardoz.txt
"myuser"@Ubuntu:~/Documents$ nano wizardoz.txt
"myuser"@Ubuntu:~/Documents$ xclip -sel c < wizardoz.txt
"myuser"@Ubuntu:~/Documents$ xclip -o -sel clip | fabric --pattern summarize
Error: Error code: 404 - {'error': {'message': 'The model `gpt-4-turbo-preview` does not exist or you do not have access to it. Learn more: https://help.openai.com/en/articles/7102672-how-can-i-access-gpt-4.', 'type': 'invalid_request_error', 'param': None, 'code': 'model_not_found'}}
Error code: 404 - {'error': {'message': 'The model `gpt-4-turbo-preview` does not exist or you do not have access to it. Learn more: https://help.openai.com/en/articles/7102672-how-can-i-access-gpt-4.', 'type': 'invalid_request_error', 'param': None, 'code': 'model_not_found'}}

**Reply by xssdoctor** [reply]
so I think you're running into an error that I introduced a few days ago but then fixed yesterday. Go into your .zshrc or .bashrc (whichever shell you're using) and delete all of the fabric aliases. Then run setup.sh again

**Reply by ValiantEagleSec** [reply]
Ok, I did that and am hitting the same error. Here's the only fabric aliases I see now in my bashrc:

alias fabric='/home/user/.cache/pypoetry/virtualenvs/fabric-b-VP_u_0-py3.10/bin/fabric'

alias fabric-api='/home/user/.cache/pypoetry/virtualenvs/fabric-b-VP_u_0-py3.10/bin/fabric-api'

alias fabric-webui='/home/user/.cache/pypoetry/virtualenvs/fabric-b-VP_u_0-py3.10/bin/fabric-webui'

alias yt='/home/user/.cache/pypoetry/virtualenvs/fabric-b-VP_u_0-py3.10/bin/yt'

**Reply by azarita1966** [reply]
I had the same error, and following advice from @xssdoctor to @ValiantEagleSec I tried to modify my .zshrc and .bashrc, but they were empty. So, no aliases to delete. 
I tried fabric --clear , but system notified that there's no such command.

So, the key here is that 'model gpt-4-turbo-preview does not exist'. I checked with fabric --listmodels and indeed this model wasn't listed. I started to search how can add this model. It turn out that you have to add more money to your balance. I went to my account at OpenAI.com, then Settings, Billing, Add to Credit Balance.you have to add minimum 5 dollars. After that the model gpt-4-turbo-preview was added to my list of models in Fabric. I run three patterns, and it deducted about 40 cents from this balance. It's on the top of monthly subscription to ChatGPT4. 
Hope that's useful. And if you have other solutions to the problem (not involving paying extra), I'd be interested.

**Comment by xssdoctor** [comment]
did you put your api key in? with --setup?

**Reply by ValiantEagleSec** [reply]
I had, I just did it again. Then I ran fabric -h and it gave me this:

fabric
usage: fabric [-h] [--text TEXT] [--copy] [--agents {trip_planner,ApiKeys}]
              [--output [OUTPUT]] [--stream] [--list] [--clear] [--update]
              [--pattern PATTERN] [--setup]
              [--changeDefaultModel CHANGEDEFAULTMODEL] [--model MODEL]
              [--listmodels] [--context]
fabric: error: argument --model/-m: expected one argument

**Reply by xssdoctor** [reply]
I have a feeling that you are looking in .zshrc and you are current in bash, or the other way around. Go into both .zshrc and .bashrc and erase all of the fabric aliases then run setup.sh

**Reply by ValiantEagleSec** [reply]
I don't think so, I just tried to open zshrc and it was a blank file

**Reply by xssdoctor** [reply]
.zshrc is a blank file? And which shell are you using?

**Reply by ValiantEagleSec** [reply]
Yep, typed "nano ~/.zshrc", it opened in the same directory that bashrc is in, but was blank and said New File. I'm using the GNOME terminal on my Ubuntu VM, so I assume it's bash.

**Reply by xssdoctor** [reply]
Oh you already did. That error message doesn’t make sense because it’s assuming you are using the —model when you aren’t so that must be one of your aliases. Do echo $shell

**Reply by ValiantEagleSec** [reply]
echo $shell

The output is blank, that seems problematic

**Comment by xssdoctor** [comment]
Ok then show me your .bashrc file

**Reply by xssdoctor** [reply]
Or maybe it’s in bash config

**Comment by majinmarco** [comment]
Hi, how can I just provide input to the model? I want to ask a question about a article I pasted. I have the following:

`pbpaste | fabric --stream` 

I want to give input to fabric like so:

`pbpaste | fabric --stream {input}` 

Obviously this does not work. How could I do it if it is possible?

**Reply by davejpeters** [reply]
You could use `echo "$(printf "Input here" && pbpaste)" | fabric -s`. This usually works for me. You can use similar methods to create an "agentic" type workflow.

**Reply by ksylvan** [reply]
Here is what you can do, @majinmarco. As an example, I highlighted your question above, copied it, then did this:

```bash
$ pbpaste | fabric --stream --search -m gpt-5.2-2025-12-11 'Can you compose a short Haiku about the question\
   being asked, then look at the repo https://github.com/danielmiessler/Fabric and try to answer it?'
```

> NOTE: Everthing that follows is the response:

Cold clipboard in pipe,  
Words seek a place to be asked—  
Fabric drinks stdin.

You don’t pass the “article text” as a positional `{input}` argument in Fabric. In Fabric’s CLI, the *input is either*:

1) **Piped in via stdin** (what you’re already doing), or  
2) **Provided as a normal trailing argument string** (for short text), while stdin is typically for longer pasted content. The README examples show stdin piping as the standard workflow. ([github.com](https://github.com/danielmiessler/fabric?utm_source=openai))

### What to do instead

#### Option A (recommended): keep using stdin (your first command is already correct)
```bash
pbpaste | fabric --stream --pattern summarize
```
(Choose whatever pattern you want with `--pattern` / `-p`.) ([github.com](https://github.com/danielmiessler/fabric?utm_source=openai))

#### Option B: “append” extra input by concatenating streams
If you want to add an extra question/instruction *after* the pasted article, concatenate it into stdin:

```bash
{ pbpaste; echo; echo "My question: What is the author's main claim?"; } \
  | fabric --stream --pattern summarize
```

Or with a variable:
```bash
question="What are the key claims and supporting evidence?"
{ pbpaste; echo; echo "Question: $question"; } \
  | fabric --stream --pattern analyze_claims
```

#### Option C: use a here-string / heredoc (no pbpaste)
```bash
fabric --stream --pattern summarize <<'EOF'
(paste article here)

Question: What are the main takeaways?
EOF
```

#### Option D: pass short input as an argument (not great for long articles)
Fabric’s README shows you can provide text after the options (example: `fabric --pattern my-analyzer "analyze this text"`). ([github.com](https://github.com/danielmiessler/Fabric?utm_source=openai))  
So for short input:
```bash
fabric --stream --pattern summarize "What does the author argue?"
```
But for an article, stdin is the right approach.

### Why `pbpaste | fabric --stream {input}` doesn’t work
`{input}` isn’t a Fabric placeholder. Fabric isn’t doing templating on the command line like that; it reads stdin as the content, and flags like `--pattern` select how to process it. ([github.com](https://github.com/danielmiessler/fabric?utm_source=openai))

**Comment by twelsh37** [comment]
Hey All. How can I submit a new pattern for possible inclusion in the patterns library for Fabric? I have created a pattern that analyses a CIPFA case study and outputs a detailed analysis of the paper, SWOT, TOWS, and Porter's Five Forces Analysis, along with a whole raft of analysis and summarisation. I created it initially to break down a paper for my wife, but I have now had time to get it all up and running, having used improve_prompt, and I would like to release it to the general public. I'm sure someone may find it helpful.

**Reply by ksylvan** [reply]
@twelsh37 Create a pull request with the new pattern and it will be merged soon afterward.

**Reply by WHITEHACK120** [reply]
I wanna learn how to code please can anyone help me with this please i need free websites for coding

**Comment by brandonp22** [comment]
Is there a way to add timestamps to some of the patterns? Ie. when doing an extract_wisdom on a 1+ hour long YT video, if the output bullets had timestamps I could go back and review that part of the video that has a quote or an idea I want to explore more.

**Reply by ksylvan** [reply]
> Is there a way to add timestamps to some of the patterns? Ie. when doing an extract_wisdom on a 1+ hour long YT video, if the output bullets had timestamps I could go back and review that part of the video that has a quote or an idea I want to explore more.

Have you tried the `youtube_summary` pattern?

```text
yt -t 'https://youtu.be/IHwt6UxiKOw?si=e77CWjeAlBs8_GJP' | fabric -p youtube_summary 
# Video Summary: Weekly Tech News Highlights

## Overview
This video covers the latest developments in technology, focusing on major announcements and partnerships involving OpenAI, Disney, and other tech giants. It also discusses advancements in AI models, industry shifts, and new product integrations.

## Key Topics and Timestamps

### GPT 5.2 Release
- **[00:00:00]** GPT 5.2 has been released, excelling in coding, knowledge work, and math. Available to paid OpenAI users.

### OpenAI and Disney Partnership
- **[00:00:20]** Disney partners with OpenAI to bring characters to Sora, allowing users to create fanfiction with Disney IP.
- **[00:01:33]** A three-year licensing agreement will enable user-generated videos featuring over 200 Disney characters, some of which will stream on Disney Plus.
- **[00:02:02]** Disney invests $1 billion in OpenAI for equity, marking a significant partnership.

### Disney's Legal Action Against Google
- **[00:02:32]** Disney issues a cease and desist to Google for alleged copyright infringement using Disney's IP in AI models.

### Time Magazine's 2025 Person of the Year
- **[00:04:14]** Time Magazine names AI architects as the 2025 Person of the Year, featuring leaders like Mark Zuckerberg and Sam Altman.

### Adobe and ChatGPT Integration
- **[00:06:41]** Adobe integrates major products like Photoshop into ChatGPT, enhancing user experience with natural language editing.

### China's DeepSeek and Nvidia Chips
- **[00:07:42]** DeepSeek reportedly uses banned Nvidia chips for AI model training, raising concerns about export violations.

### Runway ML's Gen 4.5
- **[00:09:42]** Runway ML releases Gen 4.5, a text-to-video model with impressive realism and detail.

### Meta's Shift from Open Source
- **[00:11:18]** Meta moves away from open source, focusing on monetizing AI models, with a new model codenamed Avocado.

### Rivian's Autonomy Day
- **[00:13:48]** Rivian announces updates, including custom chips and lidar integration, aiming to improve self-driving capabilities by 2026.

## Conclusion
The video highlights significant advancements and strategic moves in the tech industry, emphasizing the growing influence of AI and the importance of strategic partnerships. The developments in AI models and integrations suggest a rapidly evolving landscape with major implications for technology and media.
```
